### PR TITLE
feat: clear error on invalid node version

### PR DESCRIPTION
**Pull request by KSXGitHub**
related to https://github.com/pnpm/pnpm/issues/6909

### CODE REVIEW

File: env/node.fetcher/src/index.ts [hunk 1]
```diff
@@ -23,6 +23,17 @@ export interface FetchNodeOptions {
 }
 
 export async function fetchNode (fetch: FetchFromRegistry, version: string, targetDir: string, opts: FetchNodeOptions) {
+  // all node versions up until now (Aug 7 2023) strictly follow `vX.Y.Z` format
+  // this code should be changed if the above condition ever ceases to apply
+  if (!opts.nodeMirrorBaseUrl && !/^[0-9]+\.[0-9]+\.[0-9]+$/.test(version)) {
```
Comment (zkochan) [inline env/node.fetcher/src/index.ts]: Why should it check for the existence of nodeMirrorBaseUrl?

File: env/node.fetcher/src/index.ts [hunk 2]
```diff
@@ -23,6 +23,17 @@ export interface FetchNodeOptions {
...
+    let hint: string | undefined
+    if (/^[0-9]+\.[0-9]+$/.test(version)) {
+      hint = `Try "${version}.0"`
+    } else if (/^[0-9]+$/.test(version)) {
+      hint = `Try "${version}.0.0"`
+    }
+    throw new PnpmError('INVALID_NODE_VERSION', 'Downloadable node version must strictly satisfy the X.Y.Z syntax', { hint })
```
Comment (zkochan) [inline env/node.fetcher/src/index.ts]: Also include the used version. Something like:

```suggestion
    throw new PnpmError('INVALID_NODE_VERSION', `${version} is not an exact version. Downloadable node version must strictly satisfy the X.Y.Z syntax`, { hint })
```

You can also create a custom class for this error and move the hint creation into the constructor. Although the hint is not so important as it is already clear from the error what the issue is.

Comment (zkochan) [reply env/node.fetcher/src/index.ts]: actually, the hint is not very good as usually nobody will want the `.0.0` version but the latest instead.

Comment (KSXGitHub) [reply env/node.fetcher/src/index.ts]: > actually, the hint is not very good as usually nobody will want the `.0.0` version but the latest instead.

The hint is to give the user an idea what the exact version looks like. I didn't want to do more works than absolutely necessary (more moving parts = more chance to fail). Do you want to remove the hint instead?

File: env/node.fetcher/src/index.ts (hunk 1 above)
Comment (KSXGitHub) [reply env/node.fetcher/src/index.ts]: Other node mirrors may want to implement non-standard versions in their URL.

File: env/node.fetcher/src/index.ts (hunk 2 above)
Comment (zkochan) [reply env/node.fetcher/src/index.ts]: ok, remove it.

File: env/node.fetcher/src/index.ts (hunk 1 above)
Comment (zkochan) [reply env/node.fetcher/src/index.ts]: I don't think we should support non-standard URL formats.

Comment (KSXGitHub) [reply env/node.fetcher/src/index.ts]: OK. I will remove it.

```suggestion
  if (!/^[0-9]+\.[0-9]+\.[0-9]+$/.test(version)) {
```

**Review by zkochan — APPROVED** [review]

**Comment by zkochan** [comment]
The tests have failed.

**Comment by KSXGitHub** [comment]
> The tests have failed.

Because of `-rc`. Is there a downloadable node version with `rc`?

**Comment by KSXGitHub** [comment]
OK. It seems I have placed the validation code in the wrong place. I will switch to modifying `parseNodeEditionSpecifier`.

**Comment by KSXGitHub** [comment]
@zkochan So the CI errors was caused by `pnpm env use`. Why can `pnpm env use -g 19` succeed but `use-node-version=19` result in failure? Do you prefer `use-node-version` emit error or be the same as `pnpm env use`?

**Comment by zkochan** [comment]
`use-node-version` should work only with exact versions. `pnpm env use` can work with any specifier.

**Comment by KSXGitHub** [comment]
I have separated the parser for `.npmrc#use-node-version` from the parser for `pnpm env use`.

**Review by zkochan — APPROVED** [review]

**Pull request merged**